### PR TITLE
Switch to immutable structs

### PR DESCRIPTION
**Pull request by Tokazama**
This improves access efficiency to fields of `SimpleGraph` and `SimpleDiGraph`.
It also makes more code consistent with contributing guidelines "Minimizing use of internal struct fields".

### CODE REVIEW

**Comment by sbromberger** [comment]
Have you run benchmarks on this code? If so, could you post them?

**Comment by Tokazama** [comment]
Here's the output from this PR

```julia
  "core" => 3-element BenchmarkTools.BenchmarkGroup:
          tags: ["nv", "edges", "has_edge"]
          "nv" => 2-element BenchmarkTools.BenchmarkGroup:
                  tags: ["graphs", "digraphs"]
                  "graphs" => Trial(223.577 ns)
                  "digraphs" => Trial(222.098 ns)
          "edges" => 2-element BenchmarkTools.BenchmarkGroup:
                  tags: ["graphs", "digraphs"]
                  "graphs" => Trial(7.067 μs)
                  "digraphs" => Trial(5.874 μs)
          "has_edge" => 2-element BenchmarkTools.BenchmarkGroup:
                  tags: ["graphs", "digraphs"]
                  "graphs" => Trial(5.656 μs)
                  "digraphs" => Trial(5.706 μs)
```

and here are the benchmarks from master
```julia
1-element BenchmarkTools.BenchmarkGroup:
  tags: []
  "core" => 3-element BenchmarkTools.BenchmarkGroup:
          tags: ["nv", "edges", "has_edge"]
          "nv" => 2-element BenchmarkTools.BenchmarkGroup:
                  tags: ["graphs", "digraphs"]
                  "graphs" => Trial(231.853 ns)
                  "digraphs" => Trial(243.950 ns)
          "edges" => 2-element BenchmarkTools.BenchmarkGroup:
                  tags: ["graphs", "digraphs"]
                  "graphs" => Trial(8.197 μs)
                  "digraphs" => Trial(5.982 μs)
          "has_edge" => 2-element BenchmarkTools.BenchmarkGroup:
                  tags: ["graphs", "digraphs"]
                  "graphs" => Trial(5.587 μs)
                  "digraphs" => Trial(5.346 μs)
```

**Comment by Tokazama** [comment]
After some attempts at micro benchmarking I have

```julia
          tags: ["nv", "edges", "has_edge"]
          "nv" => 2-element BenchmarkTools.BenchmarkGroup:
                  tags: ["graphs", "digraphs"]
                  "graphs" => Trial(226.925 ns)
                  "digraphs" => Trial(208.250 ns)
          "edges" => 2-element BenchmarkTools.BenchmarkGroup:
                  tags: ["graphs", "digraphs"]
                  "graphs" => Trial(8.442 μs)
                  "digraphs" => Trial(6.003 μs)
          "has_edge" => 2-element BenchmarkTools.BenchmarkGroup:
                  tags: ["graphs", "digraphs"]
                  "graphs" => Trial(5.962 μs)
                  "digraphs" => Trial(5.062 μs)
```

I've found that the "edges" times vary pretty substantially compared to the other measures, but I assume that since immutable structs are quicker for accessing each field the great affect will be when repeatedly accessing fields of `SimpleGraph` and `SimpleDiGraph`.

File: src/operators.jl [hunk 1]
```diff
@@ -93,7 +90,14 @@ See [`reverse`](@ref) for a non-modifying version.
...
 function reverse! end
 @traitfn function reverse!(g::G::IsDirected) where G<:AbstractSimpleGraph
-    g.fadjlist, g.badjlist = g.badjlist, g.fadjlist
+    fadjlist = SimpleGraphs.fadj(g)
+    badjlist = SimpleGraphs.badj(g)
+    @inbounds for i in vertices(g)
+        f_i = copy(fadjlist[i])
+        b_i = copy(badjlist[i])
+        fadjlist[i] = b_i
+        badjlist[i] = f_i
```
Comment (piever) [inline src/operators.jl]:
```suggestion
        fadjlist[i], badjlist[i] = badjlist[i], fadjlist[i]
```

This is a pretty big performance regression though. Before `reverse!` was substantially instantaneous, whereas now it's `O(V)`. I imagine one should at least provide a fast non-mutating aliasing version.

Comment (Tokazama) [reply src/operators.jl]: Thanks for the comment. This was the one method I was worried about and I made the suggested change. One can always use `reverse` instead but I'm not sure what more can be done here.

**Comment by sbromberger** [comment]
You'll want to benchmark all the operators, as many of them rely on internal modification of the structure.

**Comment by Tokazama** [comment]
Are there any standard tests you want me to run or what it be alright to just pull from the tests?

**Comment by sbromberger** [comment]
> Are there any standard tests you want me to run or what it be alright to just pull from the tests?

Betweenness centrality is sort of my go-to benchmark for general graph accesses since it does a lot of things. But honestly, this PR will be a very heavy lift. It represents a fundamental change in the way we approach graph datastructures.

What happens to things like `add_edge!`, `rem_edge!`, etc?

Also: are you aware of StaticGraphs.jl?

File: src/operators.jl (hunk 1 above)
Comment (piever) [reply src/operators.jl]: `reverse` does a copy of the adjacency matrices as far as I understand, to avoid returning aliased graphs, so that would also not fix the performance problem.

**Comment by vlandau** [comment]
> What happens to things like `add_edge!`, `rem_edge!`, etc?

This was my question exactly

**Comment by Tokazama** [comment]
Here's a method with lots of edge changes.

```julia
#--- PR
julia> @btime transitiveclosure($(barbell))
  1.833 μs (49 allocations: 4.16 KiB)
{6, 12} directed simple Int64 graph

#--- master
julia> @btime transitiveclosure($(barbell))
  1.859 μs (49 allocations: 4.19 KiB)
{6, 12} directed simple Int64 graph


```

I can run more tests, but it shouldn't make that big of a difference here because the only difference to changing the edge is using `RefValue` (actually I might be able to improve on these benchmarks...give me a bit).

**Comment by Tokazama** [comment]
After a lot of testing I've found that this PR usually  a results in a very small improvement to performance for things that repeatedly access the `fadjlist` or `badjlist` fields, has little to no affect on methods that change edges, and tanks `reverse!`. Unless someone has an idea of how to salvage `reverse!` this is probably a bad idea.

**Comment by sbromberger** [comment]
> tanks `reverse`!

This is probably a showstopper in itself, unfortunately, but my bigger concern is that we have lots of operations that may also be impacted because they manipulate the internal fields of `SimpleGraph` directly. If we decide to keep going on this, we'll need some comprehensive benchmarks of things in `src/SimpleGraphs/generators` and `src/operators.jl`.

**Comment by Tokazama** [comment]
> my bigger concern is that we have lots of operations that may also be impacted because they manipulate the internal fields of `SimpleGraph`

Performance wise, there shouldn't be any difference when getting or setting the `ne` field, since we can inline access to the structs fields. So `add_edge!` and `rem_edge!` don't have any real difference. The real benefit here is when when calling `fadj` or `badj`, but the benefit is pretty small because usually you get those fields once and start working on the instance of `Vector{Vector{T}}` itself.

As I'm writing this out I'm realizing that it would probably be better to just ensure we get these fields at function heads to avoid repeated access to mutable fields. Sorry for waisting everybody's time. I'll close this.
